### PR TITLE
CORE-7090 Persist group parameters received from MGM

### PR DESCRIPTION
**Pull request by YashNabar**
The membership package distributed by the MGM - after a member joins, or in response to a synchronisation request - contains the updated group parameters. A member processing such synchronisation updates now persists these group parameters received from the MGM to its vnode database (and also publishes them to Kafka), on successful verification of the MGM's signature. Static registration is also updated to publish the persisted group parameters to Kafka.
https://r3-cev.atlassian.net/browse/CORE-7090

### CODE REVIEW

**Comment by YashNabar** [comment]
Target branch is set to `yash/CORE-7089` to make it easier to review. Will be retargeted to `release` once the db reconciliation stories and CORE-7089 are merged.

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipPersistenceClient.kt [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            logger.warn(this)
+            throw UnsupportedOperationException(this)
+        }
+    }
+
+    override val isRunning: Boolean
+        get() = coordinator.status == LifecycleStatus.UP
+
+    override fun start() {
+        logger.info("TestMembershipPersistenceClient starting.")
```
Comment (nikinagy) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipPersistenceClient.kt]: You can probably say something like:
`TestMembershipPersistenceClient::class.java.simpleName`

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipPersistenceClient.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        coordinator.start()
+    }
+
+    override fun stop() {
+        logger.info("TestMembershipPersistenceClient starting.")
```
Comment (nikinagy) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipPersistenceClient.kt]: and here we should stop, i guess this whole function was copy-paste from start()

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipQueryClient.kt [hunk 3]
```diff
@@ -108,12 +108,12 @@ class TestMembershipQueryClientImpl @Activate constructor(
         get() = coordinator.status == LifecycleStatus.UP
 
     override fun start() {
-        TestCryptoOpsClientImpl.logger.info("TestCryptoOpsClient starting.")
+        logger.info("TestMembershipQueryClient starting.")
```
Comment (nikinagy) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipQueryClient.kt]: TestMembershipQueryClient::class.java.simpleName

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipQueryClient.kt [hunk 4]
```diff
@@ -108,12 +108,12 @@ class TestMembershipQueryClientImpl @Activate constructor(
...
         coordinator.start()
     }
 
     override fun stop() {
-        TestCryptoOpsClientImpl.logger.info("TestCryptoOpsClient starting.")
+        logger.info("TestMembershipQueryClient starting.")
```
Comment (nikinagy) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestMembershipQueryClient.kt]: and we are stopping here too

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestCryptoOpsClient.kt [hunk 5]
```diff
@@ -212,12 +212,12 @@ class TestCryptoOpsClientImpl @Activate constructor(
         get() = coordinator.status == LifecycleStatus.UP
 
     override fun start() {
-        logger.info("TestCryptoOpsClient starting.")
+        logger.info("${this::class.java.simpleName} starting.")
```
Comment (yift-r3) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestCryptoOpsClient.kt:215]: ?

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestGroupPolicyProvider.kt [hunk 6]
```diff
@@ -57,12 +57,12 @@ class TestGroupPolicyProviderImpl @Activate constructor(
         get() = coordinator.status == LifecycleStatus.UP
 
     override fun start() {
-        logger.info("TestGroupPolicyProvider starting.")
+        logger.info("${this::class.java.simpleName} starting.")
```
Comment (yift-r3) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestGroupPolicyProvider.kt:60]: ?

File: components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt [hunk 7]
```diff
@@ -237,6 +248,19 @@ class MemberSynchronisationServiceImpl internal constructor(
                 (random.nextDouble() * 0.1 * maxDelayBetweenRequestsInMillis).toLong()
         }
 
+        private fun persistGroupParameters(viewOwningMember: HoldingIdentity, parametersFromPackage: WireGroupParameters) {
+            val groupParameters = with(parametersFromPackage) {
+                verifier.verify(
+                    mgmSignature,
+                    groupParameters.array(),
+                )
+                val parametersList = deserializer.deserialize(groupParameters.array())
```
Comment (yift-r3) [inline components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: Shouldn't we do the deserialization first?

Also, it would probably be better to keep the `array()` as a local variable instead of calling it twice.

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestCryptoOpsClient.kt (hunk 5 above)
Comment (YashNabar) [reply components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/dummy/TestCryptoOpsClient.kt:215]: Classes under that `dummy` package had several typos in the start/stop methods. This is some cleanup in line with what Niki suggested while I was touching the code there anyway.

File: components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt (hunk 7 above)
Comment (YashNabar) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: Does the order matter? We don't need the deserialised parameters during signature verification

Comment (yift-r3) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: If the byte array is corrupted, it will fail the deserialise faster than the verify (I think)...

Comment (YashNabar) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: Ah I see, changed in [7fc6619](https://github.com/corda/corda-runtime-os/pull/2519/commits/7fc6619734d4ae0d1edd8b32f6d8c71157223edb), thanks!

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/staticnetwork/StaticMemberRegistrationService.kt [hunk 8]
```diff
@@ -211,31 +214,30 @@ class StaticMemberRegistrationService @Activate constructor(
...
             } ?: this
         }
         val groupParameters = groupParametersFactory.create(groupParametersList)
 
-        // Persist group parameters for this member
-        persistenceClient.persistGroupParameters(
-            memberInfo.holdingIdentity,
-            groupParameters
-        )
+        // Persist group parameters for this member, and publish to Kafka for DB reconciliation.
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/staticnetwork/StaticMemberRegistrationService.kt]:
```suggestion
        // Persist group parameters for this member, and publish to Kafka.
```
No need to mention database reconciliation here I think.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/staticnetwork/StaticMemberRegistrationService.kt [hunk 9]
```diff
@@ -211,31 +214,30 @@ class StaticMemberRegistrationService @Activate constructor(
...
+        persistenceClient.persistGroupParameters(holdingIdentity, groupParameters)
+        groupParametersWriterService.put(holdingIdentity, groupParameters)
 
-        // If this member is a notary, persist updated group parameters for other members who have a vnode set up
+        // If this member is a notary, persist updated group parameters for other members who have a vnode set up.
+        // Also publish to Kafka for DB reconciliation.
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/staticnetwork/StaticMemberRegistrationService.kt]:
```suggestion
        // Also publish to Kafka.
```
Same here

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DistributeMembershipPackageHandler.kt [hunk 10]
```diff
@@ -123,7 +124,7 @@ class DistributeMembershipPackageHandler(
         } catch (e: Exception) {
             logger.warn("Could not distribute membership packages after registration request: '$registrationId' was approved. " +
                         "Distribution will be reattempted.", e)
-            listOf(Record(REGISTRATION_COMMAND_TOPIC, key, command))
+            listOf(Record(REGISTRATION_COMMAND_TOPIC, key, RegistrationCommand(command)))
```
Comment (charlieR3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DistributeMembershipPackageHandler.kt:127]: Why was this change required? Was it not working before?

Comment (YashNabar) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DistributeMembershipPackageHandler.kt:127]: Noticed redistribution was failing in one of the pipelines - since we cast to `RegistrationCommand` in the `RegistrationProcessor`

Comment (charlieR3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/handler/mgm/DistributeMembershipPackageHandler.kt:127]: Aha, I see this is the redistribution for error handling. I was getting confused how it worked without this change but I guess the green path distribution worked fine

File: components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt [hunk 11]
```diff
@@ -236,6 +253,24 @@ class MemberSynchronisationServiceImpl internal constructor(
                 (random.nextDouble() * 0.1 * maxDelayBetweenRequestsInMillis).toLong()
         }
 
+        private fun persistGroupParameters(
```
Comment (charlieR3) [inline components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: Nit-picky comment here - it could be good to change this to `parseGroupParameters` which takes in the membership package and remove the persistence so then where you are calling this function have:

```
val groupParameters = parseGroupParameters(viewOwningMember, updates.membershipPackage)
membershipPersistenceClient.persistGroupParameters(viewOwningMember, groupParameters)
groupParametersWriterService.put(viewOwningMember, groupParameters)
```

I think it looks a bit tidier instead of have a persist function which parses, persists, and returns group parameters to feed in to the kafka update.

Comment (YashNabar) [reply components/membership/synchronisation-impl/src/main/kotlin/net/corda/membership/impl/synchronisation/MemberSynchronisationServiceImpl.kt]: Changed in [3fee992](https://github.com/corda/corda-runtime-os/pull/2519/commits/3fee9929dcb43a0e66331aa0c74b083d3e745aca)

**Review by charlieR3 — APPROVED** [review]
LGTM

**Review by mcgovc — APPROVED** [review]
LGTM - build changes
